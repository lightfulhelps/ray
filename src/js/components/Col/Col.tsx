
import * as React from "react";
import classNames from "classnames";

const colWidths = ['xs', 'sm', 'md', 'lg', 'xl'];

type ColSizeType = boolean | number | string;

type ColumnType = boolean | number | string | {
  offset: number | string;
  order: number | string;
  size: ColSizeType;
};

type Props = {
  children: React.ReactNode;
  className?: string;
  // eslint-disable-next-line react/no-unused-prop-types
  lg?: ColumnType;
  // eslint-disable-next-line react/no-unused-prop-types
  md?: ColumnType;
  // eslint-disable-next-line react/no-unused-prop-types
  sm?: ColumnType;
  tag?: string;
  // eslint-disable-next-line react/no-unused-prop-types
  xl?: ColumnType;
  // eslint-disable-next-line react/no-unused-prop-types
  xs?: ColumnType;
};

const getColumnSizeClass = (isXs: boolean, colWidth: string, colSize: ColSizeType): string => {
  if (colSize === true || colSize === '') {
    return isXs ? 'col' : `col-${colWidth}`;
  }

  if (colSize === 'auto') {
    return isXs ? 'col-auto' : `col-${colWidth}-auto`;
  }

  if (colSize === false) return '';

  return isXs ? `col-${colSize}` : `col-${colWidth}-${colSize}`;
};

const Col = (props: Props) => {
  const {
    children,
    className,
    tag: Tag = 'div',
    ...other
  } = props;
  const colClasses = [];

  colWidths.forEach((colWidth, i) => {
    const columnProp = props[colWidth];

    if (!columnProp && columnProp !== '') {
      return;
    }

    const isXs = !i;

    if (typeof columnProp === 'object') {
      const colSizeInterfix = isXs ? '-' : `-${colWidth}-`;
      const colClass = getColumnSizeClass(isXs, colWidth, columnProp.size);

      colClasses.push(classNames({
        [colClass]: columnProp.size || columnProp.size === '',
        [`order${colSizeInterfix}${columnProp.order}`]: columnProp.order || columnProp.order === 0,
        [`offset${colSizeInterfix}${columnProp.offset}`]: columnProp.offset || columnProp.offset === 0
      }));
    } else {
      const colClass = getColumnSizeClass(isXs, colWidth, columnProp);
      colClasses.push(colClass);
    }
  });

  if (!colClasses.length) {
    colClasses.push('col');
  }

  const classes = classNames(className, colClasses);

  return <Tag {...other} className={classes}>
      {children}
    </Tag>;
};

export default Col;