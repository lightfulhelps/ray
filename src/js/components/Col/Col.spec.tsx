import React from 'react';
import { shallow } from 'enzyme';
import merge from 'lodash/merge';
import Col from './Col';

const setup = (overrides = {}) => {
  const props = merge(
    {
      children: <div>Content</div>,
    },
    overrides
  );
  const wrapper = shallow(<Col {...props} />);

  return {
    wrapper,
    props,
  };
};

describe('Col', () => {
  it('should render', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should render children', () => {
    const { wrapper } = setup({ children: 'Children' });

    expect(wrapper.text()).toBe('Children');
  });

  it('should handle the className prop', () => {
    const { wrapper } = setup();

    wrapper.setProps({ className: 'custom' });

    expect(wrapper.hasClass('col')).toBe(true);
    expect(wrapper.hasClass('custom')).toBe(true);
  });

  it('should handle all recognised col widths', () => {
    const { wrapper } = setup({ xs: 12, sm: 6, md: 4, lg: 2, xl: 1, xxl: 3 });

    expect(wrapper.hasClass('col-12')).toBe(true);
    expect(wrapper.hasClass('col-sm-6')).toBe(true);
    expect(wrapper.hasClass('col-md-4')).toBe(true);
    expect(wrapper.hasClass('col-lg-2')).toBe(true);
    expect(wrapper.hasClass('col-xl-1')).toBe(true);
    expect(wrapper.hasClass('col-xxl-3')).toBe(false);
  });

  it('should pass col size specific classes as Strings', () => {
    const { wrapper } = setup({ sm: '6' });

    expect(wrapper.hasClass('col-sm-6')).toBe(true);
    expect(wrapper.hasClass('col')).toBe(false);
  });

  it('should pass col size specific classes as Numbers', () => {
    const { wrapper } = setup({ sm: 6 });

    expect(wrapper.hasClass('col-sm-6')).toBe(true);
    expect(wrapper.hasClass('col')).toBe(false);
  });

  it('should pass col size specific classes via Objects', () => {
    const { wrapper } = setup({ sm: { size: 6, order: 2, offset: 2 } });

    expect(wrapper.hasClass('col-sm-6')).toBe(true);
    expect(wrapper.hasClass('col')).toBe(false);
    expect(wrapper.hasClass('order-sm-2')).toBe(true);
    expect(wrapper.hasClass('offset-sm-2')).toBe(true);
  });

  it('should pass col size specific classes via Objects including 0', () => {
    const { wrapper } = setup({ sm: { size: 6, order: 0, offset: 0 } });

    expect(wrapper.hasClass('col-sm-6')).toBe(true);
    expect(wrapper.hasClass('col')).toBe(false);
    expect(wrapper.hasClass('order-sm-0')).toBe(true);
    expect(wrapper.hasClass('offset-sm-0')).toBe(true);
  });

  it('should handle the tag prop', () => {
    const { wrapper } = setup();

    expect(wrapper.type()).toBe('div');

    wrapper.setProps({ tag: 'span' });

    expect(wrapper.type()).toBe('span');
  });

  it('should pass through other props', () => {
    const { wrapper } = setup({ tabIndex: 1, id: 'test' });

    expect(wrapper.prop('tabIndex')).toEqual(1);
    expect(wrapper.prop('id')).toEqual('test');
  });
});
