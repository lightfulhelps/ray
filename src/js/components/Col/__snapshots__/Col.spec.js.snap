// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col should render 1`] = `
<div
  className="col"
>
  <div>
    Content
  </div>
</div>
`;
